command/services/deregister/deregister.go

```diff
@@ -0,0 +1,114 @@
+package deregister
+
+import (
+	"flag"
+	"fmt"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/command/flags"
+	"github.com/hashicorp/consul/command/services"
+	"github.com/mitchellh/cli"
+)
+
+func New(ui cli.Ui) *cmd {
+	c := &cmd{UI: ui}
+	c.init()
+	return c
+}
+
+type cmd struct {
+	UI     cli.Ui
+	flags  *flag.FlagSet
+	http   *flags.HTTPFlags
+	help   string
+	flagId string
+}
+
+func (c *cmd) init() {
+	c.flags = flag.NewFlagSet("", flag.ContinueOnError)
+	c.flags.StringVar(&c.flagId, "id", "",
+		"ID to delete. This must not be set if arguments are given.")
+
+	c.http = &flags.HTTPFlags{}
+	flags.Merge(c.flags, c.http.ClientFlags())
+	flags.Merge(c.flags, c.http.ServerFlags())
+	c.help = flags.Usage(help, c.flags)
+}
+
+func (c *cmd) Run(args []string) int {
+	if err := c.flags.Parse(args); err != nil {
+		return 1
+	}
+
+	// Check for arg validation
+	args = c.flags.Args()
+	if len(args) == 0 && c.flagId == "" {
+		c.UI.Error("Service deregistration requires at least one argument or -id.")
+		return 1
+	} else if len(args) > 0 && c.flagId != "" {
+		c.UI.Error("Service deregistration requires arguments or -id, not both.")
+		return 1
+	}
+
+	svcs := []*api.AgentServiceRegistration{&api.AgentServiceRegistration{
+		ID: c.flagId}}
+	if len(args) > 0 {
+		var err error
+		svcs, err = services.ServicesFromFiles(args)
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Error: %s", err))
+			return 1
+		}
+	}
+
+	// Create and test the HTTP client
+	client, err := c.http.APIClient()
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error connecting to Consul agent: %s", err))
+		return 1
+	}
+
+	// Create all the services
+	for _, svc := range svcs {
+		id := svc.ID
+		if id == "" {
+			id = svc.Name
+		}
+		if id == "" {
+			continue
+		}
+
+		if err := client.Agent().ServiceDeregister(id); err != nil {
+			c.UI.Error(fmt.Sprintf("Error registering service %q: %s",
```

Comment:
Why `%q` for a string? Not a big deal just wondering.

Reply:
I think it makes for cleaner output to the user to see a quoted string, it also helps in cases where there are spaces (a service name shouldn't have spaces but just in general). So usually when erroring out to the user or even logging (w/o hclog) I use `%q` in places I'd otherwise use `%s`.